src/Encoder.cpp

```diff
@@ -0,0 +1,193 @@
+            return;
+        }
+
+        auto path = info[0].As<Napi::String>().Utf8Value();
+
+        outputFile = std::make_unique<std::ofstream>(path);
+        indexFile  = std::make_unique<zstr::ofstream>(path + ".idx");
+    }
+
+    Napi::Value Encoder::Write(const Napi::CallbackInfo& info) {
+        Napi::Env env = info.Env();
+
+        if (info.Length() == 0) {
+            Napi::TypeError::New(env, "missing argument `packet`: provide a packet to write to the file")
+                .ThrowAsJavaScriptException();
+            return env.Undefined();
+        }
+
+        Packet packet;
+        try {
+            packet = Packet::FromJsValue(info[0], env);
+        }
+        catch (const std::exception& ex) {
+            Napi::TypeError::New(env, std::string("invalid type for argument `packet`: ") + ex.what())
+                .ThrowAsJavaScriptException();
+            return env.Undefined();
+        }
+
+        // Calculate the NBS Packet's full size
+        uint32_t size = WritePacket(packet);
+        WriteIndex(packet, size);
+
+        bytesWritten += size;
+
+        return this->GetBytesWritten(info);
+    }
+
+    Napi::Value Encoder::GetBytesWritten(const Napi::CallbackInfo& info) {
+        return Napi::BigInt::New(info.Env(), this->bytesWritten);
+    }
+
+    void Encoder::Close(const Napi::CallbackInfo& info) {
+        if (this->IsOpen(info).As<Napi::Boolean>()) {
+            this->outputFile.get()->close();
+            this->indexFile.get()->close();
+        }
+    }
+
+    Napi::Value Encoder::IsOpen(const Napi::CallbackInfo& info) {
+        return Napi::Boolean::New(info.Env(), this->outputFile.get()->is_open());
+    }
+
+    uint64_t Encoder::WritePacket(const Packet& packet) {
+
+        // NBS File Format
+        // Name      | Type               |  Description
+        // ------------------------------------------------------------
+        // header    | char[3]            | NBS packet header ☢ { 0xE2, 0x98, 0xA2 }
+        // length    | uint32_t           | Length of this packet after this value
+        // timestamp | uint64_t           | Timestamp the data was emitted in microseconds
+        // hash      | uint64_t           | the 64bit hash for the payload type
+        // payload   | char[length - 16]  | the data payload
+
+        // The size of our output timestamp, hash and data
+        uint32_t size = sizeof(packet.timestamp) + sizeof(packet.type) + packet.length;
+
+        uint64_t timestampMicros = packet.timestamp / 1000;
+
+#pragma pack(push, 1)
+        struct PacketHeader {
+            std::array<char, 3> header = {char(0xE2), char(0x98), char(0xA2)};
+            uint32_t size{0};
+            uint64_t timestamp{0};
+            uint64_t hash{0};
+
+            PacketHeader(const uint32_t& size, const uint64_t& timestamp, const uint64_t& hash)
+                : size(size), timestamp(timestamp), hash(hash){};
+        };
+#pragma pack(pop)
```

Comment:
Also when you have types like these, define them at the top of the file not in the function (same for the other types too)

```suggestion
#pragma pack(push, 1)
        struct PacketHeader {
            std::array<char, 3> header = {char(0xE2), char(0x98), char(0xA2)};
            uint32_t size{0};
            uint64_t timestamp{0};
            uint64_t hash{0};

            PacketHeader(const uint32_t& size, const uint64_t& timestamp, const uint64_t& hash)
                : size(size), timestamp(timestamp), hash(hash){};
        } __attribute__((packed));
```